core/src/main/java/org/apache/brooklyn/core/objs/BasicSpecParameter.java

```diff
@@ -404,7 +431,126 @@ public static void addParameters(AbstractBrooklynObjectSpec<?, ?> spec, List<? e
             spec.parametersAdd(BasicSpecParameter.fromSpec(loader.getManagementContext(), spec));
         }
         if (explicitParams.size() > 0) {
-            spec.parametersAdd(explicitParams);
+            spec.parametersReplace(resolveParameters(explicitParams, spec.getParameters(), spec));
+        }
+    }
+
+    /** merge parameters against other parameters and known and type-inherited config keys */
+    static Collection<SpecParameter<?>> resolveParameters(Collection<? extends SpecParameter<?>> newParams, Collection<? extends SpecParameter<?>> existingReferenceParamsToKeep, AbstractBrooklynObjectSpec<?,?> spec) {
+        Map<String,SpecParameter<?>> existingToKeep = MutableMap.of();
+        if (existingReferenceParamsToKeep!=null) {
+            for (SpecParameter<?> p: existingReferenceParamsToKeep) { 
+                existingToKeep.put(p.getConfigKey().getName(), p);
+            }
+        }
+
+        List<SpecParameter<?>> result = MutableList.<SpecParameter<?>>of();
+        for (SpecParameter<?> p: newParams) {
+            if (p instanceof SpecParameterForInheritance) {
+                SpecParameter<?> existingP = existingToKeep.remove(p.getConfigKey().getName());
+                if (existingP!=null) {
+                    p = ((SpecParameterForInheritance<?>)p).resolveWithAncestor(existingP);
+                } else {
+                    // TODO find config keys on the type (not available as parameters)
+                    /* we don't currently do this due to low priority; all it means if there is a config key in java,
+                     * and a user wishes to expose it as a parameter, they have to redeclare everything;
+                     * nothing from the config key in java will be inherited */
+                    p = ((SpecParameterForInheritance<?>)p).resolveWithAncestor((ConfigKey<?>)null);
+                }
+                result.add(p);
+            } else {
+                existingToKeep.remove(p.getConfigKey().getName());
+                result.add(p);
+            }
+        }
+        result.addAll(existingToKeep.values());
+        return result;
+    }
+
+    /** instance of {@link SpecParameter} which includes information about what fields are explicitly set,
+     * for use with a subsequent merge */
+    @SuppressWarnings("serial")
+    @Beta
+    static class SpecParameterForInheritance<T> extends BasicSpecParameter<T> {
```

Comment:
the separate class is to know when something still needs to be inherited and when inheritance has been completed.  would welcome a second opinion here @neykov because it is ugly, but i think it should never leak ie the ugliness is entirely hidden.